src/Build/Logging/BaseConsoleLogger.cs

```diff
@@ -911,6 +891,195 @@ public int Compare(object o1, object o2)
             }
         }
 
+        /// <summary>
+        /// Helper class to indent all the lines of a potentially multi-line string with
+        /// minimal CPU and memory overhead.
+        /// </summary>
+        /// <remarks>
+        /// <see cref="IndentString"/> is a functional replacement for the following code:
+        /// <code>
+        ///     string IndentString(string s, int indent)
+        ///     {
+        ///         string[] newLines = { "\r\n", "\n" };
+        ///         string[] subStrings = s.Split(newLines, StringSplitOptions.None);
+        ///     
+        ///         StringBuilder result = new StringBuilder(
+        ///             (subStrings.Length * indent) +
+        ///             (subStrings.Length * Environment.NewLine.Length) +
+        ///             s.Length);
+        ///     
+        ///         for (int i = 0; i &lt; subStrings.Length; i++)
+        ///         {
+        ///             result.Append(' ', indent).Append(subStrings[i]);
+        ///             result.AppendLine();
+        ///         }
+        ///     
+        ///         return result.ToString();
+        ///     }
+        /// </code>
+        /// On net472, benchmarks show that the optimized version runs in about 50-60% of the time
+        /// and has about 15% of the memory overhead of the code that it replaces.
+        /// <para>
+        /// On net7.0 (which has more optimizations than net472), the optimized version runs in
+        /// about 70-75% of the time and has about 30% of the memory overhead of the code that it
+        /// replaces.
+        /// </para>
+        /// </remarks>
+        private static class OptimizedStringIndenter
+        {
+#nullable enable
+            internal static string IndentString(string? s, int indent)
+            {
+                if (s is null)
+                {
+                    return string.Empty;
+                }
+
+                string newLine = Environment.NewLine;
+
+                using PooledSpan<StringSegment> segments = GetStringSegments(s);
+                int indentedStringLength = ComputeIndentedStringLength(segments, newLine, indent);
+                string indented = new string('\0', indentedStringLength);
+
+                unsafe
+                {
+#pragma warning disable SA1519 // Braces should not be omitted from multi-line child statement
+                    fixed (char* pInput = s)
+                    fixed (char* pIndented = indented)
```

Comment:
When targeting .NET Core, you can avoid unsafe code and use string.Create. That also saves writing nulls into the string. Not sure whether conditional compilation is worthwhile though

https://learn.microsoft.com/en-us/dotnet/api/system.string.create?view=net-7.0#system-string-create-1(system-int32-0-system-buffers-spanaction((system-char-0)))

Reply:
Now using `String.Create` on .NET Core and a cached `StringBuilder` on .NET Framework.